helix-term/src/commands/typed.rs

```diff
@@ -1410,47 +1410,51 @@ fn tutor(
     Ok(())
 }
 
+fn abort_goto_line_number_preview(cx: &mut compositor::Context) {
+    if let Some(last_selection) = &cx.editor.last_selection {
+        let scrolloff = cx.editor.config().scrolloff;
+        let (view, doc) = current!(cx.editor);
+        doc.set_selection(view.id, last_selection.clone());
+        view.ensure_cursor_in_view(doc, scrolloff);
+        cx.editor.last_selection = None;
+    }
+}
+
 pub(super) fn goto_line_number(
     cx: &mut compositor::Context,
     args: &[Cow<str>],
     event: PromptEvent,
 ) -> anyhow::Result<()> {
     match event {
-        PromptEvent::Abort => {
-            if let Some(line_number) = cx.editor.last_line_number {
-                goto_line_impl(cx.editor, NonZeroUsize::new(line_number));
-                let (view, doc) = current!(cx.editor);
-                view.ensure_cursor_in_view(doc, line_number);
-                cx.editor.last_line_number = None;
-            }
-            return Ok(());
-        }
+        PromptEvent::Abort => abort_goto_line_number_preview(cx),
         PromptEvent::Validate => {
             ensure!(!args.is_empty(), "Line number required");
-            cx.editor.last_line_number = None;
```

Comment:
I am pretty sure that `PromptEvent::Validate` can happen immidietly without reciving a `PromptEvent::Update`. In that case the command does nothing now wheras it worked correctly before. I think you can restructure the funciton slightly to account for that. First check if we are aborting => call `abort_goto_line_number_preview` and return.

Then run the code that is currently in the update branch.

The check if we are validating and run the code here. `last_selection` can then be obtained with `cx.editor.last_selection.clone().unwrap()` (as it's always initalized a few lines prior)

Reply:
> I am pretty sure that PromptEvent::Validate can happen immidietly without reciving a PromptEvent::Update.

How would that happen? Keybinding?

Reply:
I think keybindings may do that. Haven't tested it but all other typed commands work when imminently called with ` PromptEvent::Validate` so I wouldn't want to diverge from that here.

Reply:
Re-reading your comment:

> First check if we are aborting => call abort_goto_line_number_preview and return.

What do you mean by "first check"? The function only handles one event at a time, and it does call the abort function and return then an abort event is fired.

Reply:
Right now you do a match on the event:
``` rust 
match event{
PromptEvent::Abort => ..,
...
}
```

Instead use a simple if:

```
if event == PromptEvent::Abort{
// abort ...
return
}

// update selection

if event == PromptEvent::Validate{
// jumplist entry and clear last_selection
}
```

Reply:
Sorry, but I don't understand the difference, both constructs result in the function returning after calling abort, no?

Reply:
The difference is that you don't need to copy all the update code into the validate branch.

With a match the full thing would be:
``` rust 
match event{
PromptEvent::Abort => /* abort */,
PromptEvent::Update => /* update selection  */,
PromptEvent::Validate => /* update selection, add jumplist entry and clear last_selection */,
...
}
```

I suggested the structure above so you can share the update code between the `Update` and the `Validate` branch

Reply:
Okay, I see what you're saying. I'm trying to figure out the case where validate is called with arguments without update having been called. If I can find that, I'll restructure the code accordingly.

Reply:
@pascalkuthe You were, of course, exactly right -- binding a key in insert mode to `:goto 123` sends only the Validate event, which currently doesn't work. New commit addresses that. Thanks for catching that and the detailed explanation!